### ISSUE TITLE
Using mongeez to ensure indexes fails

### ISSUE THREAD

**Issue by stoffer** [mongeez/mongeez#18]
I'm trying to setup and use mongeez, as it seems like just the tool to use for mongodb data migration management.

And inserting data also works like a charm, but I would like to ensure indexes on certain keys in some of my collections. Mongeez adds a new document in the mogeez collection saying the changelog has been run, but now index is added to the collection.

If I run the mongo script for ensuring index directly in the mongo shell it works fine.

Any idea or comments ?

Thanks
Christoffer Magnussen

**Comment by notdang** [comment]
Happens for me too. Apparently db.eval acquires the write/read global locks on the DB and the background:true tries to re-index in background, which is not possible at this stage.

**Comment by pamelaluis** [comment]
Yes, I've observed this issue too. The problem I've found is that mongeez uses db.eval(script) which results in a global write lock and somehow this conflicts with creating indexes in the background.

The error you would usually get if you were to run the following command on the mongo shell:
db.eval("db.<collectionName>.ensureIndex({fields:1},{background:true});")

error: "can't start bg index b/c in recursive lock (db.eval?)"

A fix for this would probably be to workaround this issue by using db.command() and passing the unlock:true option. Sample code that appears to resolve this problem is below:

           DBObject command = BasicDBObjectBuilder.start()
                .add("$eval", code)
                .add("nolock", true)
                .get();
            CommandResult retVal = db.command(command);
            retVal.throwOnError();
            Object commandReturnValue = retVal.get("retval");

            if (commandReturnValue != null && DBObject.class
                .isAssignableFrom(commandReturnValue.getClass())) {
                DBObject dbObjectRetVal = (DBObject) commandReturnValue;
                Object errorMessage = dbObjectRetVal.get("errmsg");
                if (errorMessage != null) {
                    throw new CommandFailureException(retVal);
                } else {
                    LOG.info("Command: " + code);
                    LOG.info(commandReturnValue.toString());
                }
            }

**Comment by WeipingGuo** [comment]
any update on this issue? i would like to create an index on existing big data set and would like to do background indexing. thanks.
